### ISSUE TITLE
Update UK Admin Boundary Polygons

### ISSUE THREAD

**Issue by vicchi** [whosonfirst-data/whosonfirst-data#1871]
See right at the very end of this issue for background ...

## Rationale

The UK admin boundary polygons are somewhat venerable and a little bit ... _shonky_ in places, with a lot of them originating from Quattroshapes. This issue is the start of a proposal to refresh and update UK admin geographies and to keep them up to date, or at least more up to date-er than they currently are.

## Data Sources

Probably the most official source for this information is the [UK Boundaries Commission](https://boundarycommissionforengland.independent.gov.uk/) who, unhelpfully, don't actually _release_ any data. Instead, data is contributed directly to Ordnance Survey, who then surface the data in a range of data products. From an open data perspective, this means [OS Boundary-Line](https://osdatahub.os.uk/downloads/open/BoundaryLine).

But ... from experience working with this data set, whilst it's probably the most _authoritative_ it's _not_ the easiest or best to work with. Particularly where coastal boundaries sometimes extrude to capture close lying offshore islands and sometimes ... _don't_. There doesn't seem to be any method of scheme behind this.

Instead, the [UK Office for National Statistics (ONS)](https://www.ons.gov.uk/), take OS Boundary-Line, clean it up, and align it with the administrative hierarchy of the UK in all of its weird historical gloriousness, assign consistent identifiers (AKA GSS codes). See also https://github.com/whosonfirst/whosonfirst-properties/issues/100, which is related to this.

See also this diagram which I drew up to try and make sense of the UK's admin geographies together with the relevant GSS code prefixes.

ONS boundary data is in two flavours, basically one which follows the high water mark of the coastline and one which follows the low water mark. This is grossly over-simplifying but will suffice for this discussion.

## Proposal

This is not entirely altruistic. For my day job I'm starting to make heavy of use WOF, both via Pelias, and also on its own. It will make by day job life easier if the UK has better admin coverage and accuracy. But also it's good to do this, especially given both my relationship and also @straup's with WOE as a precursor to WOF. Pause for a moment for misty eyed nostalgia of the days of Flickr and Yahoo! Geo ...

So I propose to update the WOF GeoJSON files with the relevant and applicable ONS boundary data ...

## Licensing

ONS data is released and licensed under the UK's Open Government License (OGL) and data from ONS, under this license, has already been imported into WOF as has OS open data under the same license. See also #1685 which covers ONS postal codes (ONS PD) from @tomtaylor.

## Coverage

Despite this being for the UK ... it's not. Because politics and history and probably stupidity. This will mostly cover England, Wales and Scotland (to a greater degree) but not Northern Ireland. Because Ordnance Survey and Ordnance Survey Northern Ireland are two different organisations who don't always cooperate as we'd like them to. Because politics. Again. And also because Northern Ireland's Land and Property Services. There is _some_ NI open data, but not as much and because ONS don't have authority for this part of the UK.

If this all works out as planned in my head, there may be a subsequent proposal for NI data, but let's not get too far ahead ...

## Approach

1. Build a local WOF in a box style instance on a beefy local server so I can poke and spelunk on the current state of WOF data in general and for the UK in particular.
1. Import the applicable ONS boundary data into something geospatial, either ES or PostGIS (I've mostly already done both of these)
1. Work out a mapping between the UK GSS geographies and the applicable WOF placetypes
1. Try and derive and automated way to determine the concordance between WOF IDs and GSS codes
1. Update the WOF GeoJSON with the ONS data
1. Try and derive some regression tests, particularly around overlaps and ensuring boundaries are contiguous where it makes sense for them to be so
1. Pull request!

I doubt it will be that simple, but as a brain dump of an approach this will probably suffice for tonight.

Any and all comments of suggestions, particularly around the approach from people who have probably done this sort of thing before would be _really_ welcome.

## Background

From a conversation in the `pelias/pelias` Gitter channel ...

>I wonder if some of this might be down to it looking like the WOF admin hierarchy puts this address not in the neighbourhood of Teddington (whosonfirst:neighbourhood:85790893) but in the neighbourhood of Strawberry Hill (whosonfirst:neighbourhood:85790805) which is very close to Teddington but this specific address is most definitely not in that neighbourhood.

> A quick visual inspection shows that the boundary polygon for Strawberry Hill (https://spelunker.whosonfirst.org/id/85790805/) has made a land grab for a big chunk of what constitutes Teddington (https://spelunker.whosonfirst.org/id/85790893/)

> Looks like these polygons came from Mapzen and from QuattroShapes … so I’m wondering if I tried to adjust the WOF admin polygons with those sourced from a UK specific open data set, such as ONS or OS, and then rebuild Pelias with these, would that help this?

Then, in the `mapzen/whosonfirst` Gitter channel ...

> @thisisaaronland @tomtaylor @nvkelso I assume wrangling the UK admin polygons would be a “good thing” for WOF in general?

>Better than Quattroshapes at least

> @straup :- @vicchi Given how weird and squiggly UK admin boundaries are I think any help would be welcome - as always the principal issue is licensing

> @thisisaaronland The data is all under the OGL as a lot of OS open data is. There’s already been OGL licensed data inserted into WOF so I assume there’s no issue there

> Sorry. To be complete - ONS and OS open data is under the OGL. Unless it’s not. In which case I won’t.

> @nvkelso ... @vicchi Yes, those shapes are a little wonky. A lot has changed in UK in the last 8 years, and for open data great wins :) Please coordinate with @stepps00  by filing a GitHub issue so we can document and clear the source and discuss the approach and progress. We've already imported some other open UK data from those sources so should be good.

**Comment by vicchi** [comment]
Updated with admin diagram which I forgot to attach. Duh

**Comment by tomtaylor** [comment]
Hi @vicchi! 👋 I am definitely in favour of this - would be great to get the UK in better shape.

**Comment by missinglink** [comment]
Potentially related to this https://github.com/whosonfirst-data/whosonfirst-data/issues/1792 which discusses the 'higher level' places and could do with a review from someone who lives there ;)

**Comment by straup** [comment]
@vicchi I think there is value in your approach, if only to understand better where the stress points are.

Having a methodology, not to mention a toolchain, for defining the rules by which "authoritative" datasets can be managed and merged in to WOF would be useful.

**Comment by vicchi** [comment]
@straup @tomtaylor @stepps00 @missinglink @nvkelso @orangejulius ... so I've been busy ...

With the caveat of "_it's complicated because history, geography, politics and devolved governmen_t" I've tried to fit the "not really hierarchical in places" geography of the United Kingdom into a "_sort of hierarchy with lots of edge cases_" hierarchy.

This is also a first pass at working out how UK admin geography _might_ fit into the current WOF placetype model and also mapping the Office for National Statistics GSS code system to the different ONS data sets and WOF placetypes. In some cases it works surprisingly well. In other cases, almost all in England, it's not so much fitting a square peg into a round hole, it's more smashing a large dodecahedron into a small round hole. With a sledgehammer.

It's also linked I think to #1792?

This is also by no means authoritative; it's simplified a lot of the rough edges and there are many and varied edge cases, but it's a start and it looks something like this ...

![image](https://user-images.githubusercontent.com/442617/89818624-f9e34600-db41-11ea-9cec-2f7177a04354.png)

So ... the United Kingdom is simple enough. Mostly. Lacking a GSS code (though a convention on a pseudo-code exists), the UK is a WOF `country`.

Scotland, Wales and Northern Ireland are countries, though NI is often termed province, but these exist in and map to a WOF `macroregion`. These three countries also have devolved government, which might explain why their administrative hierarchy is simple and straightforwards.

Again, I'm using ONS terms here and other terms of reference do exist.

Scotland is divided into 32 council areas. There's enough history in place to argue that council areas should sit at the same level in the hierarchy as English _regions_, but also that they could sit at the same level as English _counties_ and given the way in which they're governmentally administered and are comprised of historical counties, so that's where I've put them, as a WOF `county`. That's as low level as the admin hierarchy gets. The next level down is towns and villages, so all Scottish towns/villages should be classed as a WOF `locality` and there isn't a concept of a WOF `localadmin` in Scotland per se.

The same simple approach applies in Wales, with 22 counties/regions/local government areas/unitary authorities (depending on which source you look at). As with Scotland, I've classed these as a WOF `county` and there's no `localadmin` in Wales. Also, don't confuse mentions of Welsh unitary authorities and English unitary authorities; the name is the same, the concept most definitely isn't.

Ditto for Northern Ireland, with 11 local government districts, ditto for classing them as a WOF `county` and ditto for no WOF `localadmin`.

And then England, where it all goes a bit crazy.

England has 9 regions, some of which exist in WOF as a `region`, some of which don't. It seems to make sense to keep each of these as a WOF `region`, though some tidying up is needed with other WOF places and where they should go in the placetype model.

Next is a group of admin areas which I've lumped in as a `WOF` county, some of which are, some of which aren't. So English counties, combined authorities, unitary authorities and metropolitan counties (which are defunct but still exist in law and as a widely used reference term). There's a subtle parent child relationship here, which seems to defy a rules based system, but at a high level, combined authorities are parent to (some) counties and also to (some) unitary authorities, which in turn parent (some) metropolitan counties.

_I have no idea how to handle this right now._

Greater London also fits in at a county level, sort of. The City of London probably also fits in here, though it might also fit at the next level down, which is ...

Boroughs and other districts. London Boroughs are direct children of Greater London. Metropolitan Districts are direct children of (Metropolitan) Counties. Non Metropolitan Districts are direct children of Counties, _not_ Unitary Authorities or Combined Authorities. These all seem to fit as a WOF `borough`.

Almost there ... 

There's also historic and ceremonial counties. The former are defunct, such as Middlesex, but are still widely used in addresses, though ignored by the formal Royal Mail address standard. The latter, still exist but are as the name suggests, ceremonial in nature. However ... these also map almost completely to the old historic postal counties of the UK, as for Middlesex, and are valuable contexts for geocoding and address parsing in the UK.

WOE used to have the concept of historical placetypes, but I _also have no idea how, if at all, this could be handled or represented in WOF_.

Finally, there's post towns, which _are_ part of the Royal Mail address standard. There's ~1,500 of these. Some of which match to existing WOF `localadmin` places, some to existing WOF `locality` places and some not at all. But for UK geocoding, these are _really_ important and a swine to get boundaries for unless you pay for Royal Mail licensed data. I'll need to somehow manually curate these. Joy.

So ... for everywhere apart from England, there's a rules based system, though quite a lot of editorialising with be needed to shunt things around where WOF places already exist in this proposed set of mappings.

For England, I think it's going to be a brute force approach and a manual and tedious one to set up per place mappings between existing WOF documents and where they might live in the grand new scheme of things. But without proprietary data, I don't really see any other alternative and while this is going to be complex, at least the ONS data (plus some OSNI data for, err, Northern Ireland) is well maintained and regularly updated.

Oh ... and this _doesn't_ (yet) cover the Isle of Man, Channel Islands, Crown Dependencies and so on.

Help, suggestions, poking and prodding the model and/or telling me where I'm wrong is needed please.

**Comment by vicchi** [comment]
So where I wrote _I have no idea how to handle this right now_ ... after thinking some more I have _some_ suggestions on how to handle these, though they may not be workable or even right.

## Historic Counties

Officially no longer in use, most were killed off in 1965, but despite this they are still widely in use as an unofficial part of addresses. Officially, they're not part of the Royal Mail address standard. But they're still _in_ addresses and, especially for geocoding (read: Pelias), they need to be recognised even if only to discount them.

There's an OS Open Data set for these, which helps.

But how best to represent something historic but which remains in colloquial use? We _could_ create a new place type but I assume that would have a big impact on lots of code and the assumptions that code makes.

These counties could fit into the hierarchy via a secondary hash in the `wof:hierarchy` property, and also in some cases where there's a one to one "mapping" between the old county and what came after, via the `wof:coterminous` property too.

As these counties are ceased but not deprecated, we could possibly be clever with EDTF dates, having `edtf:inception` and `edtf:cessation` represent the start and end of their "official" life, but _not_ having `edtf:deprecated` ... and/or `is_current` as `0`, `is_ceased` as `1` and `is_deprecated`/`is_superseded`/etc as `0` ... ?

## Ceremonial Counties

There's 48 of these since 1997, again these can be modelled as a WOF `county` placetype with a secondary `wof:hierarchy` hash and also, in places, `wof:coterminous`. There's also an OS open data set for these as well

## Counties, Combined Authorities and other Horrors

>There's a subtle parent child relationship here, which seems to defy a rules based system, but at a high level, combined authorities are parent to (some) counties and also to (some) unitary authorities, which in turn parent (some) metropolitan counties.

Again, I think that secondary hierarchies and coterminous flags would work here.

## Post Towns

The geographic make up of each of these is well defined in terms of the postcode. Either the area code or the area and sector code which make up each post town is in the public domain and documented. I _may_ be able to infer the polygon from the published and open data centroid of each postcode, to get a fuzzy boundary polygon which might be enough. But this needs to be tested and established if this is accurate enough.

Place type wise, these _feel_ like a `locality` albeit an alternate one in the hierarchy (as above). They're not administrative per se and are composed _of_ postcodes, so `locality` seems to fit?

**Comment by nvkelso** [comment]
> Post Towns

We've settled on having `postalcode` records carry the pseudo locality name instead of overloading the true meaning of locality. See https://github.com/whosonfirst-data/whosonfirst-data/issues/202#issuecomment-409306352 and later comments.

_Still digesting the rest..._

**Comment by vicchi** [comment]
@nvkelso The starting point was to use `pelias/lastline` as a work around for this, but then some (but not all) post towns don't have a one to one analogue with some other `localadmin` or `locality` place.

So I'm struggling to see how this would work in the hierarchy as, for example, postcodes starting `TW11` are part of the _post town_ of Teddington, whose boundary polygon is not equivalent to the boundary of the `locality` which represents Teddington in WOF.

So how would tagging every postcode in `TW11` with a post-town-esque psuedo locality name work? Assuming the postcode _place_ in WOF has a centroid, could I also add a boundary polygon as well (though that would probably need upstream code changes to recognise this in, say, Pelias).

In the UK post towns are a thing and it feels like ignoring them from a more formal hierarchy side steps the issue that they're a critical part of the UK address, albeit a part which doesn't fit into the formal admin hierarchy.

With apologies for having my "UK geocoding problems" hat on here!

**Comment by thisisaaronland** [comment]
The UK is a good "What's going in there anyway..." use-case to filter and test things against. I'm sure other places could make things as interesting as the UK but that shouldn't take away from what the UK has accomplished ;-)

A couple quick comments:

* `edtf:cessation` represents when something stopped being "common currency", for lack of a better expression.  `etdf:deprecated` represents when a record was deemed to be "beyond repair".

The definition of what constitutes "current" will vary from place to place so there aren't formal rules but flagging historic counties as "deprecated" would be a mistake. If they are still being used, even in a vernacular sense, then there doesn't need to be any reason to flag them as "ceased" either. It probably makes sense to note that the ONS isn't using them but that can be done with a `ons:` property or properties.

* Postal codes are assumed to have multiple hierarchies or more specifically to span multiple descendants of the same place type.

**Comment by vicchi** [comment]
In the spirit of _learning_ by doing, I've manually built a real world example, not falling into the usual trap of "my house as a good example", but a rather fine riverside pub in Cornwall I may have visited a few times.

Turns out it's a good example of how one physical location can have multiple addresses which don't agree with each other.

* The pub's web site says their address is `Pandora Inn, Restronguet Creek, Mylor Bridge, Falmouth, Cornwall, TR11 5ST`
* Google Maps says the address is `Pandora Inn, Restronguet Hill, Creek, Falmouth, TR11 5ST`
* Royal Mail says the address is `Pandora Inn, Restronguet, Mylor, Falmouth, TR11 5ST`

Before we get into place types and hierarchies, this disagreement in address is actually extremely pertinent.

Royal Mail defines the address components as:

* Pandora Inn - business
* Restronguet - road name (which is incorrect, the road name is either Restronguet Hill _or_ Quay Road)
* Mylor - dependent locality (Mylor is both a village around 2 miles to the west and a Civic Parish)
* Falmouth - post town (not to be confused with Falmouth the town, but see below)
* TR11 5ST - postcode

the venue defines the address components as:

* Pandora Inn - business
* Restronguet Creek - not a road name but actually the part of the river Fal that the venue is on the bank of
* Mylor Bridge  - dependent locality (Mylor Bridge is also a village around 1 mile to the west)
* Falmouth - post town (not to be confused with Falmouth the town which is around 5 miles away, but see below)
* Cornwall - historic postal county
* TR11 5ST - postcode

I'll ignore Google Maps except to note that I can't find _any_ reference to Creek as a place name in any other data source.

Based on ONS and OS Open Data (and also including a UPRN from a proprietary OS data set), the hierarchy looks something like this ...

* `K02000001`: United Kingdom (source: ONS)
* `E92000001`: England (source: ONS)
* `E12000008`: South West (Region) (source: ONS)
* `?????????`: Cornwall (Ceremonial County) (source: OS BoundaryLine)
* `?????????`: Cornwall (Historic County) (source: OS BoundaryLine)
* `E06000052`: Cornwall (Unitary Authority) (source: ONS)
* `E04011494`: Mylor (Civil Parish) (source: ONS)
* `?????????`: TR11 5ST (source: ONS)
* `10090524624`: Pandora Inn (source: OS Proprietary)

... where the identifiers which commence `^[A-Z][0-9]` are ONS GSS codes and `?????????` is no GSS code exists. With the exception of the missing post town this looks pretty good

Comparing this with what's currently in WOF for the venue ...

* `wof:id` - [`85633159`](https://spelunker.whosonfirst.org/id/85633159/): United Kingdom (`country`)
* `wof:id` - [`85684547`](https://spelunker.whosonfirst.org/id/85684547/): Cornwall (`region`) (superseded by [`1360699085`](https://spelunker.whosonfirst.org/id/1360699085/))
* `wof:id` - [`135390911`](https://spelunker.whosonfirst.org/id/135390911/): Pandora Inn (`venue`)

... and in WOF for the postcode ...

* `wof:id` - [`85633159`](https://spelunker.whosonfirst.org/id/85633159/): United Kingdom (`country`)
* `wof:id` - [`404227469`](https://spelunker.whosonfirst.org/id/404227469/): England (`macroregion`)
* `wof:id` - [`85684547`](https://spelunker.whosonfirst.org/id/85684547/): Cornwall (`region`) (superseded by 1360699085)
* `wof:id` - [`404449471`](https://spelunker.whosonfirst.org/id/404449471/): Mylor (`localadmin`)
* `wof:id` - [`487910229`](https://spelunker.whosonfirst.org/id/487910229/): TR11 5ST (`postalcode`)

Firstly the hierarchies are old, incomplete in places and need a refresh, even before we look at updating the data.

But that aside, a more complete WOF hierarchy could look like this

* `wof:id` - [`85633159`](https://spelunker.whosonfirst.org/id/85633159/): United Kingdom (`country`)
* `wof:id` -  [`404227469`](https://spelunker.whosonfirst.org/id/404227469/): England (`macroregion`)
* `wof:id` - `XXXXXXXXX`: South West (`region`) (new WOF place derived from the ONS South West region)
* `wof:id` - [`1360699085`](https://spelunker.whosonfirst.org/id/1360699085/): Cornwall (`county` or `macrocounty`) (updated boundary from the ONS Cornwall Unitary Authority)
* `wof:id` -` XXXXXXXXX`: Cornwall (`county`) (new WOF place derived from OS Historic Counties)
* `wof:id` - `XXXXXXXXX`: Falmouth (`localadmin` or `macrohood`) (new WOF place for the postal town)
* `wof:id` - [`404449471`](https://spelunker.whosonfirst.org/id/404449471/): Mylor (localadmin) (updated boundary from the ONS Civil Parish)
* `wof:id` - [`487910229`](https://spelunker.whosonfirst.org/id/487910229/): TR11 5ST (postalcode)
* `wof:id` - [`135390911`](https://spelunker.whosonfirst.org/id/135390911/): Pandora Inn (venue)

Also of note - each of these places would, where appropriate, have a regenerated hierarchy _and_ a concordance for `ons:gss` for the ONS GSS code.

Also also of note, WOF currently has two Falmouth places

* Falmouth as a `localadmin` (https://spelunker.whosonfirst.org/id/404438245/) which looks like it's either the current Falmouth Town Council area (post 1974) or the Falmouth Borough Council area (pre 1974), but it's difficult to say
* Falmouth as a `locality (https://spelunker.whosonfirst.org/id/101852391/) whose polygon _looks_ identical to the other Falmouth and is either the town itself _or_ the Falmouth Town Council area. There's no metadata in WOF I can find to say _which_ sort of Falmouth this is other than the placetype.

@nvkelso Just to keep the discussion on the postal town semantic going, the comment from @stepps00 on #202 would potentially work for having the postcode's preferred name as its post town, however for the purposes of tokenising a full or partial address based on WOF I don't know whether this will work or whether a `macrohood` would be a better place type. This is because there's a one to many relationship of postowns to (dependent) postcodes and also the post town itself forms part of the address string in addition to and to qualify the postcode itself.

@thisisaaronland For historic counties as a `county` placetype ... how would we manage the historic version versus the current version. In some cases there is no current county per se, in some cases there is and the boundaries overlap or align and in some cases they're completely different. Would an alternate hierarchy hash suffice? What other metadata property could be used to say, for example, "this is the historic Cornwall"?

Also I'm conveniently ignoring the concept of UK ceremonial counties as they're little used and this is complex and has enough moving parts as it is. Filed under _another day perhaps_ for now.

Finally for now .. it should be obvious that `XXXXXXX` above entails the creation of a new WOF _place_. I assume the values for a new ``wof:id`` come from Brooklyn, Mission or London Integers? But that's the correct process and tooling for doing this?

**Comment by straup** [comment]
"...how would we manage the historic version versus the current version (of a county)"

This is where the `wof:supersedes` and `wof:superseded_by` stuff comes in to play. The historic version never goes away, it is simply superseded by a contemporary record.

You might now be thinking: That means cascading updates to all the descendants of a record when it's superseded, doesn't it? The answer is: Yes. It's an operational challenge but, in 2020, this is the work.

The canonical example has been the Olympic Stadium in Sarajevo in 1984, 1993 and 2020: The physical structure is the same but the context is not the same in meaningful and quantifiable ways. Or, terminals and boarding areas at an airport... 

* https://whosonfirst.org/blog/2016/10/06/wof-lifecycle-documentation/
* https://whosonfirst.org/blog/2017/06/29/tackling-space-and-time-in-whosonfirst/
* https://mw19.mwconf.org/paper/mapping-space-time-and-the-collection-at-sfo-museum/

The goal is to a) figure out how to automate all of this (see also @tomtaylor 's work with UK postcodes) and b) how to annotate things so that there are suitable flags and signals for purpose-fit applications like geocoding or map tiles without painting the data in the corner demanded by any one application's needs.

As far as place types go it sounds like we've arrived at a concrete manifestation of something we've said about hierarchies in WOF:

* That any record can have as complex a hierarchy as befits that place; the only hard rule is that every place, across the entire corpus, has at least one entry in a common hierarchy.

The canonical example for place types and hierarchies is New York City: If you and I have two different applications using WOF the rule is that we both have all the `common` place types, which includes `locality`, in our respective databases. But we may differ on the need to index the `borough` place type (aka "Brooklyn" or "Manhattan"). We may not be able to find commonality but at least we can agree on "New York City".

I have been pretty strident about not being liberal in adding place types and I think that's contributed to some of confusion about how to tackle the UK (and places like it) within the current hierarchy of place types.

FWIW, this is something we bump up against routinely at SFO Museum trying to shoehorn a museum collection and airport-related things in to the WOF model (see link above). That experience has gotten me to thinking that maybe what WOF needs is a catch-all `custom` place type where semantics and rules stored in properties external to the hierarchy.

I haven't thought about this for a while so it's a bit fuzzy and I am not sure how it could be done in a way that would allow tools to automatically deference things; for example, all place types and sources have machine-readable metadata that can be resolved and processed given their unique ID or label.

The point I am trying to make, without necessarily having a workable solution yet, is that the goal isn't to squeeze all of the UK in to a single hierarchy of place types. If the UK, or other places, have genuinely unique groupings then there should be some way to honour them. The cost of that specificity is that those places are special snowflakes and have to be accounted for as such in any application that uses them. Again, the details of how that would work are fuzzy at best right now.

The value of using the core place types (`common`, `common_optional` and `optional`) is that it preserves a common measure of "slicing and dicing" functionality for the entire planet and leaves purpose-fit filtering to the bespoke properties.

The convention is to use Brooklyn Integers for minting new WOF IDs. The `py-mapzen-whosonfirst-export` package takes care of generating IDs for records that don't have them.

There is also this which was written to help @tomtaylor with the process of seeding a bazillion new IDs for the UK postcodes:

https://github.com/whosonfirst/go-whosonfirst-id-proxy

It is currently targeted at being included in custom Go code but I can write a pair of simple tools to expose the same functionality as a command line tool and HTTP pony.

**Comment by vicchi** [comment]
Will read and fully digest later but for now ..

> The point I am trying to make, without necessarily having a workable solution yet, is that the goal isn't to squeeze all of the UK in to a single hierarchy of place types

I think multiple hashes in the `wof:hierarchy` will allow this to be modelled. I think as well as honouring this ..

> That any record can have as complex a hierarchy as befits that place; the only hard rule is that every place, across the entire corpus, has at least one entry in a common hierarchy.

Also ...

> I have been pretty strident about not being liberal in adding place types and I think that's contributed to some of confusion about how to tackle the UK (and places like it) within the current hierarchy of place types.

The more I think about this the more I agree and we _should_ be able to model this with a combination of additional hierarchy hashes without additional placetypes. 

I _am_ still confused and fuzzy about how we can model post towns, but that seems to be the final hurdle (@nvkelso, @tomtaylor  and @stepps00, please pitch in on this one if you can and help me work it out in my head and in a working example).

This is all on the assumption that the other elements of the working example in my previous comment _feel_ like the right direction of travel. So yes, no, maybe or _huh_?

As with anything like this, the devil will be in the details and I'm sure we'll find gotchas along the way, but I want to try and get the principle behind this nailed down as much as would be humanly and realistically possible before I launch into a jouyous spree of scripting and coding stuff.

**Comment by vicchi** [comment]
@straup So based on this ...

>You might now be thinking: That means cascading updates to all the descendants of a record when it's superseded, doesn't it? The answer is: Yes. It's an operational challenge but, in 2020, this is the work.

I assume this means there isn't any tooling currently to

* impute a hierarchy given a centroid, so a sort of point in polygon query tool that say "find me all documents with a boundary polygon that this point falls into" ... if not, then I'm thinking either an ES `geo_polygon` query might help, or I go full on and push all the `wof:id` and geometry polygons from the `spelunker` index _into_ PostGIS and use `ST_Within`/`ST_Contains`, which would be more work. Again, this is assuming such a tool doesn't exist.

* cascadingly update the descendents as you mention above

... ?

It does feel that there's good progress being made here in understanding all of this but some of the subtleties and moving parts are still eluding me. Timezone differences notwithstanding, perhaps we could jump on a video conferencing tool of choice and just thrash this out once and for all?

**Comment by thisisaaronland** [comment]
There are parts of the tooling but like a lot of things they are incomplete. Specifically:

* There is no single monolithic spatial index for the entire dataset. This is largely driven by economic concerns.

* There are tools for indexing slices of the data spatially, notably https://github.com/whosonfirst/go-whosonfirst-pgis and https://github.com/whosonfirst/go-whosonfirst-mysql (MySQL 5.7+ has proper spatial support). There was a corresponding SQLite + spatialite package but that seems to have stopped working because of the changes to SQLite, spatialite or both. The principal advantage of SQLite is that you don't have deal with administering a database server. The principal disadvantage (of SQLite) is that wrangling spatialite often makes administering a database server seem easy by comparison. Then there is the spatial stuff in ES which was premature during the Mapzen days and you might be the first person to put it through its paces. To that end there is https://github.com/sfomuseum/go-whosonfirst-elasticsearch which has hooks for indexing geometries.

* Beyond that it's not always a straightforward (interects or contains) spatial problem. A good example is to ask what continent Russia is part of? At Mapzen we tried to separate the logic to interpret the results of a spatial query from the spatial query itself. That is, some times math (geometry) doesn't win because "humans". For example the https://github.com/whosonfirst/py-mapzen-whosonfirst-hierarchy and https://github.com/whosonfirst/py-mapzen-whosonfirst-spatial packages. The latter honours math, the former honours WOF.

* The `py-mz-wof-hierarchy` package is currently the single most authoritative place for the logic governing how and where hierarchies get rebuilt, for example handling the `wof:controlled` property which is an explicit flag signaling to automated processes that a human has interceded in the metadata for a record and they should not try to apply their rule-set to it.

* `py-mz-wof-hierarchy` has not been updated to Python 3 but I can do that soon.

* The rules that govern `py-mz-wof-hierarchy` are all, literally, encoded in code. That's something that should be written down in sentences and paragraphs as a rule set that can be translated to other software tools.

* We did have some early-stage tools for applying cascading updates for a record (all the neighbourhoods in a locality, etc.) but I am hard-pressed to remember where they are right now and those tools would probably benefit from a fresh start anyway.

* But yeah, given two "parent" records (let's say a locality) you would need to find 3 sets of records: Descendants that are contained by both the old and new records; Anything not contained by that overlap in the respective old and new records. For that point you could walk the contents of each list creating new records with the relevant parent, possibly looking it up spatially if a (new) descendant is no longer contained by the new parent, and then building the hierarchies according to the WOF rules and updating the `wof:supersedes` and `wof:superseded_by` properties accordingly.

A video chat would be lovely. Let's follow up by email.

**Comment by straup** [comment]
FYI: https://github.com/whosonfirst/py-mapzen-whosonfirst-hierarchy/releases/tag/v2.0.0

**Comment by vicchi** [comment]
> py-mz-wof-hierarchy has not been updated to Python 3 but I can do that soon.

If I get to needing that and you lack bandwidth, I'll give it a crack. I am no Python expert but slow process on bringing WOE back to life (yes, that's still going on in the background) has given me enough Python-fu to at bring some P2 scripts into the P3 fold. Heck, I've even written some (WOE) modules which seem to work and install and don't crash.

> A video chat would be lovely. Let's follow up by email.

Email to follow ...

**Comment by nvkelso** [comment]
@vicchi hows this one been going?

**Comment by vicchi** [comment]
@nvkelso @straup @tomtaylor @thisisaaronland It's been a (long) while, but I'm now revisiting this. Specifically gathering together all the UK official open data boundaries and their identifiers, pushing them into PostGIS (as a side effect of work I've been doing for the day job) and squinting a lot to draw up a concordance mapping between WOF and how the UK (government) sees itself.

It's complex, but we know this already, so to try and make this somewhat manageable, I've sliced and diced the UK into two main hierarchies; yellow is the official admin hierarchy and green is the official electoral hierarchy. Both hierarchies cross paths several times (epecially electoral wards which live in both hierarchies) but all places in the hierarchies are notionally parented by one of the countries that make up the United Kingdom.

![UK Admin and Electoral Hierarchies](https://user-images.githubusercontent.com/442617/210138064-105d5b38-bb2b-4760-9be3-0fb7b97d73ff.png)

So starting at the UK level ...

The UK maps to a WOF country; at least currently, there's possibly a discussion to be had as to whether the UK is a country or an empire and thus whether England, Northern Ireland, Scotland and Wales are macroregions or countries. But I'll just seed that thought for now and keep things aligned with WOF as it currenty stands.

England, Northern Ireland, Scotland and Wales are thus macroregions (but see the previous paragraph).

From hereon this is a proposal and open to kicking around ...

Administratively ...

The England only Region maps to a WOF region

Counties, Metropolitan Counties, Combined Authorities and the Greater London Authority map to a WOF county - officially there's only ~24 counties but thematically these are all county-like and only exist in England

Unitary Authorities, Non-metropolitan and Metropolitan Districts, London Boroughs, Welsh Unitary Authorities (not to be confused with their English equivalents), Scottish Council Areas and Northern Ireland Local Government Districts _probably_ map to a WOF localadmin

Electoral Wards live in both the admin and electoral hierarchy and map to a ... WOF neighbourhood? macrohood? microhood? New electoral ish place type? Dunno.

And finally Welsh Communities and English Parishes (of both flavours) map to ... something thematically the same size-ish as wards but different, (shrugs emoji)

Electorally ...

All constituencies are notionally parented by their country. There's national hierarchies of Westminster constituencies and their wards (the same ones as in the admin hierarchy). There's the London Assembly constuencies, notionally parented by the GLA and notionally childed (not a real word) by the wards that make up the London Boroughs.

There's the Welsh and Scottish devolved parliaments and their consituencies and regions (which are analogous to wards).

And there's the Northern Ireland Assembly areas and constituencies which are confusing the hell out of me and which I need to research more and try and get boundaries for the constituencies or at least draw them up manually.

Now if that's not a good way to end the year .... yeah. Maybe.

Thoughts?

**Comment by nvkelso** [comment]
> England, Northern Ireland, Scotland and Wales are thus macroregions (but see the previous paragraph).

One solution we'e used elsewhere would be to set a [placetype_local](https://github.com/whosonfirst/whosonfirst-properties/tree/main/properties/wof#placetype_local) of `Country` on  them, eg:

-  https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/blob/master/data/404/227/471/404227471.geojson

**Comment by nvkelso** [comment]
And wither the Ceremonial Counties? WOE had them as "colloquial areas" IIRC?

**Comment by thisisaaronland** [comment]
So, apparently I've thought about custom placetype specifications enough to write code to merge them with "core" in the Go code:

https://github.com/whosonfirst/go-whosonfirst-placetypes/blob/main/specification.go#L108

@vicchi If you can make a set of UK constituency-specific placetype definitions I can try to plug them in to Go tools and that would give us something to "poke" at.

https://github.com/whosonfirst/whosonfirst-placetypes

**Comment by tomtaylor** [comment]
Something else to fry in your mind woks: I think postcode regions would be useful, e.g. W1A, M60. These are often used colloquially and refer to the area bounded by postalcodes starting with that prefix. They would parent postalcodes and by parented by the country/macroregion, I expect.

Happy to have a specific discussion about these somewhere else, but thought it was worth sharing if we’re thinking about new placetypes in the UK.

**Comment by thisisaaronland** [comment]
In Canadia, these are referred to as "forward sorting areas". For example `T6R`:

https://raw.githubusercontent.com/whosonfirst-data/whosonfirst-data-postalcode-ca/master/data/504/804/689/504804689.geojson

Is the FSA for `T6R 2R3`:

https://raw.githubusercontent.com/whosonfirst-data/whosonfirst-data-postalcode-ca/master/data/554/865/797/554865797.geojson

These are both parented by the locality of Edmonton rather than the latter by the former. I think that was just done out of expediency at the time but it makes a certain amount of sense.

Specifically, the idea that `wof:parent_id` should refer to a "core" placetype (common, optional or common optional). That's never been explicitly stated anywhere but I think it's a reasonable requirement. Having said that now, SFO Museum is probably violating that rule...

**Comment by nvkelso** [comment]
Focusing just on administrative geography, @vicchi @tomtaylor can you take this for a spin? Sorry we've been talking about this for _years_ now...

The TL;DR:

- add missing "Government Region" (12 count) as `region` 
- push the existing "Ceremonial County" (102) from `region` to `macrocounty` (an unused placetype)
- consider consolidating `localadmin` "Civil Parish" / "Community" (11k) to `locality` instead (where it is often x-listed in `placetype_alt`)
- this would open up `localadmin` for the missing "Local Authority Districts" (361), which are sometimes coterminous with their parents, and in Northern Ireland are the lowest level gov't and same for Scotland communities
- NOTE: There is some work already at the WOF `borough` level for "London Borough" and "Metropolitan District" features but worth another look...

I think all the other administrative geography is already accounted for in WOF? I'm deliberately ignoring both electoral and postal geographies in this issue as they are being discussed elsewhere and require new placetypes / new tooling.

level | WOF placetype (current) | WOF placetype (proposed) | placetype_local | count | names | note
----- | ------------------------ | --------------------------- | ---------------- | ------ | ------ | ----
-1 | `empire` | _no change_ | ??? | 1 | | Includes home country and British Overseas Territories, including extant "crown dependencies", "sovereign bases" & etc
0 | `country` | _no change_| "Country" ? | 1 | Cleanup names per https://github.com/whosonfirst-data/whosonfirst-data/issues/1792  | Limited to just "Great Britain and Northern Ireland"
1 | `macroregion` | _no change_ | "Constituent Country" | 4 | Cleanup names per https://github.com/whosonfirst-data/whosonfirst-data/issues/1792 |  England, Wales, Scotland, Northern Ireland
2a | _new_ | `region` | "Government Region" | 12 |  | _Missing features, propose adding in Sept 2023._
2b | `region` | `macrocounty` | "Ceremonial County" | 102 | | "Historic counties" expired in 1965. They can be added also at this placetype, as long as they have a cessation date set so they don't interfere with current set. `See also`: https://github.com/whosonfirst-data/whosonfirst-data/pull/1368.
3 | `macrocounty` | _revised_ | | | | | _Unused as of 2023-08-29, propose shifting "Ceremonial Counties" here in Sept 2023._
4 | `county` | _no change_ | "Administrative county", "Non-metropolitan District", and "Unitary Authority" | 173 | | @vicchi notes these might better be labeled "Combined Authority", "Greater London Authority", "Metropolitan County", "County"
5a | `localadmin` | _possible new_ | "Local Authority Districts" (including "Community" in Scotland) | 361 | | These generally are lowest level of functional local government
5b | `localadmin` | _possible change_ | "Civil Parish" or "Community" | 11,400 + some number of locality as placetype_alt | | Gaps in Shapefile in urban areas to be fixed in https://github.com/whosonfirst-data/whosonfirst-data/issues/2152. These are mostly advisory rather than actual local gov't.
6 | `locality` | _possible change_ | _Varies, unofficial_ | ~ 16k with 1/3 of those being points | | Most locality points are in Scotland and Northern Ireland (@stepps00 I think you cleaned up all GB in https://github.com/whosonfirst-data/whosonfirst-data/pull/1436, so remaining points are for larger admin areas that span many many individual localities and match WOF expectations)

**Comment by nvkelso** [comment]
Someday I'll meet you at this example pub @vicchi, resurrecting the abbreviated example:

The pub's web site says their address is `Pandora Inn, Restronguet Creek, Mylor Bridge, Falmouth, Cornwall, TR11 5ST`

Royal Mail defines the address components as:

* Pandora Inn - business
* Restronguet - road name (which is incorrect, the road name is either Restronguet Hill _or_ Quay Road)
* Mylor - dependent locality (Mylor is both a village around 2 miles to the west and a Civic Parish)
* Falmouth - post town (not to be confused with Falmouth the town, but see below)
* TR11 5ST - postcode

the venue defines the address components as:

* Pandora Inn - business
* Restronguet Creek - not a road name but actually the part of the river Fal that the venue is on the bank of
* Mylor Bridge  - dependent locality (Mylor Bridge is also a village around 1 mile to the west)
* Falmouth - post town (not to be confused with Falmouth the town which is around 5 miles away, but see below)
* Cornwall - historic postal county
* TR11 5ST - postcode

Based on ONS and OS Open Data (and also including a UPRN from a proprietary OS data set), the hierarchy looks something like this ...

* `K02000001`: United Kingdom (source: ONS)
* `E92000001`: England (source: ONS)
* `E12000008`: South West (Region) (source: ONS)
* `?????????`: Cornwall (Ceremonial County) (source: OS BoundaryLine) –– WOF does have this
* `?????????`: Cornwall (Historic County) (source: OS BoundaryLine) –– WOF would only have this as ceased `macrocounty`
* `E06000052`: Cornwall (Unitary Authority) (source: ONS)
* `E04011494`: Mylor (Civil Parish) (source: ONS)
* `?????????`: TR11 5ST (source: ONS)
* `10090524624`: Pandora Inn (source: OS Proprietary)

... where the identifiers which commence `^[A-Z][0-9]` are ONS GSS codes and `?????????` is no GSS code exists. With the exception of the missing post town this looks pretty good

A more complete WOF hierarchy could look like this

* `wof:id` - [`85633159`](https://spelunker.whosonfirst.org/id/85633159/): United Kingdom (`country`)
* `wof:id` -  [`404227469`](https://spelunker.whosonfirst.org/id/404227469/): England (`macroregion`)
* `wof:id` - `XXXXXXXXX`: South West (`region`) (new WOF `region` place proposed for Sept 2023 derived from the ONS South West region)
* `wof:id` - [`1360699085`](https://spelunker.whosonfirst.org/id/1360699085/): Cornwall (`region` now, but proposed as `macrocounty` for Sept 2023) (updated boundary from the ONS Cornwall Unitary Authority)
* `wof:id` -` XXXXXXXXX`: Cornwall (`macrocounty`) (new but ceased WOF `macrocounty` place derived from OS Historic Counties)
* `wof:id` -` XXXXXXXXX`: YYYYY (`localadmin`) (new WOF place derived from OS Local Authority Districts theme)
* `wof:id` - `XXXXXXXXX`: Falmouth (`locality`point only missing the nearby real town point...)
* `wof:id` - [`404449471`](https://spelunker.whosonfirst.org/id/404449471/): Mylor (localadmin) (updated boundary from the ONS Civil Parish)
* `wof:id` - [`487910229`](https://spelunker.whosonfirst.org/id/487910229/): TR11 5ST (postalcode should get a new property for `Falmouth` to accomplish "postal locality")
* `wof:id` - [`135390911`](https://spelunker.whosonfirst.org/id/135390911/): Pandora Inn (venue)

**Comment by thisisaaronland** [comment]
Fun fact: This pub is not found in the Overture Data places dataset.

```
> grep -e 'Pandora' overture-GB.jsonl | grep Cornwall
{ "type": "Feature", "properties": { "id": "tmp_DE1A3BC01FA46E209854FB299E65E926", "updatetime": "2023-07-24T00:00:00.000", "version": 0, "confidence": 0.9202608108520508, "websites": [ "http:\/\/www.pandora.net\/en-gb\/stores\/pandora-east-kilbride" ], "social": [ "https:\/\/www.facebook.com\/119257345382582" ], "emails": null, "brand": { "names": { "brand_names_common": [ { "value": "Pandora", "language": "local" } ] }, "wikidata": null }, "addresses": [ { "locality": "East Kilbride", "postcode": "G74 1LW", "freeform": "300 Cornwall Street East Kilbride Shopping Centre", "country": "GB" } ], "categories": { "main": "jewelry_store", "alternate": [ "shopping", "fashion_accessories_store" ] }, "sources": [ { "dataset": "meta", "property": "", "recordid": "119257345382582" }, { "dataset": "msft", "property": "\/properties\/existence", "recordid": "562949958279203" } ] }, "geometry": { "type": "Point", "coordinates": [ -4.1797453, 55.7603197 ] } }
```

**Comment by nvkelso** [comment]
**Mid October 2023 progress update and import plan**

First PR for **United Kingdom** coarse admin:

- Existing `macroregion` features (4) 
    - Updated geometries from OS (default) and OSNI (default and rev-geo alt geom)
    - No property updates
    - Concordances: 
        - `gbr-os:census_code` for England, Wales, and Scotland.
        - Will research the Northern Ireland code. 
        - Note, these seem different then the existing `gbr-ons:gss_code` we have, determine which one is "right"
- New `region` features (186)
    - All existing `county` features (plus 1 new one for a split) get duplicated here, and the old `region` move to `macrocounty`
    - IDs on duplicated features change because of lifecycle rules around changing placetypes, but they will get superced{ed} properties to link them to the still current `county` features.
    - These features function as the admin-1 and are reported that way to ISO. But rant: This matches ISO except for the London boroughs, which WOF will only track as `county`.
    - Updated geometries from OS and OSNI
    - New properties from
        - OS (name, area_description, area_type_description, census_code)
        - OSNI (lgdname, lgdcode)
    - Concordances: 
        - OS (census_code)
        - OSNI (lgdcode)
    - These are a mix of "administrative counties" and the "unitary authority" and/or "single-tier" unitary authorities / communities in England, Wales, and Scotland... and LGAs in Northern Ireland.
    - If we wanted to, we could bring in the other "historic" administrative counties here as mz:is_current = 0 with cessation date (this is the most WOF thing to do). Or we could replace the lieutenency areas in the `macrocounty` slot with "historic" counties?!
- New `macrocounty` - _was an unused slot_
     - The existing "lieutenancy areas" aka "ceremonial counties" (102) move from WOF `region` to `macrocounty`
     - See below for discussion about these areas and "historical counties"... It's kinda odd to have a non-administrative unit in here that doesn't nest with the rest and is more about the monarchy than civil administration... But I'm not British so I'm probably missing something...
    - The IDs change because of lifecycle rules around changing placetypes, but they will get superced{ed} properties to link them.
    - Geometries: we _could_ import new geoms if we wanted to from OS and modified version of OSNI data.
    - Properties: There aren't new properties to bring in.
    - Concordances: These features don't have concordances as they relate to the sovereign not demographic administration or statistics. 
- New `county` (361) records as they are the admin-2 of UK
    - New "divisions" (197) of the now WOF `region` "administrative counties" (22) are new features, so new IDs. This will include features for the London boroughs, which will need to be marked coterminous and hierarchy_label 0.
    - The old "administrative counties" (22) get marked not current (replaced by "divisions"), superceded into new WOF `region` features.
    - Remaining 164 existing "single-tier" unitary authority WOF `county` features are coterminous features with the moved `region` features, no change to IDs. But they should also marked hierarchy_label = 0, but sharing any concordances and properties with the `region` for demographic mapping. We keep them as separate features (versus setting a placetype_alt on the `region` for `county) since `county` is a common instead of common-optional placetype.
    - All together, these changes unblock demographic mapping at this placetype
    - Updated geometries from OS and OSNI
    - New properties from
        - OS (name, area_description, area_type_description, census_code)
        - OSNI (lgdname, lgdcode)
    - Concordances: 
        - OS (census_code)
        - OSNI (lgdcode)

Second PR

- **Northern Ireland** `locality` and `neighbourhood`
    - We only had 7 `locality` polygons and ~650 `locality` points
    - Using OSNI planning settlement limit polygons, this yields... (an existing WOF source)
    - 337 of the locality points gain polgyons (and adjusted label centroids from OSNI), the 7 existing polygons get cleaned up, 325 new locality polygons (with OSNI label centroids) are imported, 25 points get shifted around, and 25 new points are added. Overall +8%.
    - There were 87 neighbourhood points before (with human curated label positions)
    - Around 10 neighbourhood points / polgyons are adjusted. No new neighbourhood points to speak of.
    - Updated geometry will note if it's raw `osni` geom source or modified "src_via" `whosonfirst`.
    - New properties will be imported in `osni` namespace (name, latitude, longitude; plan_name, location, lgd_id, lgd_name, comments)
    - Concordances are not great here. There a settlement polygon GUID we could import, but it doesn't actually look like a concordance ID or code. The new settlement points don't have GUID or codes.
    - This work ignores "townlands" which are something between subdivisions and townships in the USA and useful for genealogy. There is open data for this level available both in Northern Ireland and Ireland, but it seems totally irrelevant / paved over in urban areas, and somewhat relevant in rural areas. Future work might consider adding "townlands" for `localadmin` import as a way of backfilling locality names everywhere in rev-geo. 
- **Scotland** `locality` and `neighbourhood`
    - We had 20 locality polygons and 4297 locality points
    - 494 locality points gain polygons (and the 20 polygons get updated, too), 171 new locality polygons, 1599 points shift to a better location, 2510 new locality points added. Overall +60%.
    - Most of these are straight geometry import, some have modified geometry and should be noted as such during import
    - We had several hundred neighbourhoods (mostly polygon) concentrated in the two big cities. 
    - Some of the neighbourhood polygons we have now outside of London are super wonky QS popcorn, they should at least be clipped to their parent.
    - 96 neighbourhoods (~50%) those could get a better label position, and 1,227 new "suburb" neighbourhood points will be imported (6x)
    - Geometry updates and new geoms from new `gbr-nrs` source... and modifications noted as `whosonfirst` with `src_via` note.
    - Properties prefixed with 
        - `gbr-nrs` (code, name, lbl_lng, lbl_lat)
        - `os` ( id, name1, name1_lang, name2, name2_lang, type, local_type, most_detail_view, least_detail_view, mbr_xmin, mbr_xmax, mbr_ymin, mbr_ymax, postcode_district, country, same_as_dbpedia, same_as_geonames)
    - Concordances are possible here: 
        - Scotish `gbr-nrs:code` (eg: `S19002084`) -- _this one is most like the existing gbr-ons concordances..._
        - OS: `os:id` (eg: `osgb4000000074335040`) -- the web links dropped the `osgb` part, but their linked data web site is offline

Third PR (work still in progress):

- `localtiies` in **England** and **Wales**
    - Around 11k almost exclusively locality polygons – which are mostly coterminous with the localadmin
    - Better open data has come out from OS, we should reset the coterminous locality to points and then compare with newer settlement polygons and points. Its OK (weird as that is) for a locality to span multiple `localadmin` as the civil parish are more advisory in nature, and the parent `county` has the ultimate say. Rarely the OS lets a settlement polygon cross a unitary authority boundary (nea `county).
    - Will switch over to mix of settlement polygons (7.7k) and OS open name populated place points (~27.0k) or +3x
    - Our earlier work to name (and sometimes split) the "FILLER" / "unparished" areas of the urban cores will be preserved at this level
    - Geometry updates and new geoms from new `gbr-os` source... and modifications noted as `whosonfirst` with `src_via` note.
    - Properties prefixed with 
        - `os` ( id, name1, name1_lang, name2, name2_lang, type, local_type, most_detail_view, least_detail_view, mbr_xmin, mbr_xmax, mbr_ymin, mbr_ymax, postcode_district, country, same_as_dbpedia, same_as_geonames)
        - `os` (gsscode, name1_text, name1_language, name2_text, name2_language)
    - Concordances are possible here: 
        - OS: `os:id` (eg: `osgb4000000074335040`) -- the web links dropped the `osgb` part, but their linked data web site is offline
        - OS: `os:gsscode` (eg: `E63001403`) -- most like existing codes
- `localadmin` in **England** and **Wales**
    - Will get a refresh for latest "civil parish" polygons (around 11k old, around 11k new, ±0%)
    - Since "FILLER" / "unparished" areas don't have any useful properties, we will probably not import them, or we import as the derived named locality polygons but mark them as statistical gores and hierarchy_label = 0?
    - This placetype level remains important for rural "locality" name fallbacks for rev-geo
    - Same geom, properties, and concordances as `locality`
- `neighbourhood` in **England** and **Wales**
    - We have 1.1k neighbourhood polygons now, and 11.9k neighbourhood points
    - Will conflate OS open name populated place "suburb" points (~9.4k) with the above, there will be small % new features.
    - Some of the neighbourhood polygons we have now outside of London are super wonky QS popcorn, they should at least be clipped to their parent.
    - Same geom, properties, and concordances as `locality`
- `borough` in **England** and **Wales**
    - Special care for **London** (coterminous with localadmin and county) and **Cardiff** (coterminous with localadmin) features, please
    - Same geom, properties, and concordances as `locality`

Sources: 

- OS (existing WOF source)
- OSNI (existing WOF source)
- Scotland (new WOF source)
    - `gbr-nrs`
    - National Records of Scotland 
    - https://www.nrscotland.gov.uk/statistics-and-data/geography/our-products/settlements-and-localities-dataset/settlements-and-localities-digital-boundaries

& etc

- Postal counties (from the address example) are not "historic" Administrative Counties  or "ceremonial" counties: https://en.wikipedia.org/wiki/Postal_counties_of_the_United_Kingdom
    - Just like "locality_name" is an optional property for WOF `postcode` features
    - A new "county_name" could be added as an optional property for WOF `postcode` features. But RoyalMail hasn't maintained those for 10 to 20 years (because they were confusing and because the forward sort 3-4 char component solves that generically now).

@tomtaylor @vicchi @stepps00 @thisisaaronland @missinglink

**Comment by nvkelso** [comment]
First PR for coarse admin hierarchy updates in United Kingdom is up for review: 

- https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/pull/92
